File: onebusaway-android/src/main/java/org/onebusaway/android/speed/ScheduleSpeedEstimator.java
/*
 * Copyright (C) 2024 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.speed;

import org.onebusaway.android.io.elements.ObaTripSchedule;

/**
 * Estimates speed using the trip schedule: finds the two stops bracketing the vehicle's
 * current scheduled position and computes segment speed from the timetable.
 */
public class ScheduleSpeedEstimator implements SpeedEstimator {

    @Override
    public Double estimateSpeed(String vehicleId, VehicleState state, VehicleSpeedTracker tracker) {
        if (state.getScheduledDistanceAlongTrip() == null) {
            return null;
        }

        ObaTripSchedule schedule = tracker.getSchedule(state.getActiveTripId());
        if (schedule == null) {
            return null;
        }

        ObaTripSchedule.StopTime[] stopTimes = schedule.getStopTimes();
        if (stopTimes == null || stopTimes.length < 2) {
            return null;
        }

        double currentDist = state.getScheduledDistanceAlongTrip();

        // Find the two stops bracketing the current position
        int beforeIdx = -1;
        int afterIdx = -1;

        for (int i = 0; i < stopTimes.length; i++) {
            if (stopTimes[i].getDistanceAlongTrip() <= currentDist) {
                beforeIdx = i;
            } else {
                afterIdx = i;
                break;
            }
        }

        // Edge cases: before first stop or after last stop
        if (beforeIdx == -1) {
            // Before the first stop - use first segment speed
            beforeIdx = 0;
            afterIdx = 1;
        } else if (afterIdx == -1) {
            // After the last stop - use last segment speed
            beforeIdx = stopTimes.length - 2;
            afterIdx = stopTimes.length - 1;
        }

        double distDelta = stopTimes[afterIdx].getDistanceAlongTrip()
                - stopTimes[beforeIdx].getDistanceAlongTrip();
        long timeDelta = stopTimes[afterIdx].getArrivalTime()
                - stopTimes[beforeIdx].getDepartureTime();

        if (distDelta <= 0 || timeDelta <= 0) {
            return null;
        }

        return distDelta / timeDelta;
    }
}

File: onebusaway-android/src/main/java/org/onebusaway/android/extrapolation/math/speed/ScheduleSpeedEstimator.kt
/*
 * Copyright (C) 2024-2026 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.extrapolation.math.speed

import org.onebusaway.android.extrapolation.data.TripDataManager
import org.onebusaway.android.extrapolation.math.DiracDistribution
import org.onebusaway.android.io.elements.ObaTripSchedule
import org.onebusaway.android.io.elements.ObaTripStatus
import org.onebusaway.android.io.elements.speedAtDistance

/**
 * Estimates speed using the trip schedule: finds the two stops bracketing the vehicle's current
 * scheduled position and computes segment speed from the timetable.
 */
class ScheduleSpeedEstimator(private val dataManager: TripDataManager) : SpeedEstimator {

    override fun estimateSpeed(tripId: String, queryTime: Long): SpeedEstimateResult {
        val status = dataManager.getLastState(tripId)
        val schedule = dataManager.getSchedule(tripId)
        return estimateFromStatus(status, schedule, queryTime)
    }

    override fun estimateSpeed(tripId: String, queryTime: Long,
                               snapshot: TripDataManager.TripSnapshot): SpeedEstimateResult =
            estimateFromStatus(snapshot.lastState, snapshot.schedule, queryTime)

    private fun estimateFromStatus(
            status: ObaTripStatus?,
            schedule: ObaTripSchedule?,
            queryTime: Long): SpeedEstimateResult {
        if (status == null) return SpeedEstimateResult.Failure(
                SpeedEstimateError.InsufficientData("No state for trip"))

        if (queryTime < status.lastLocationUpdateTime) {
            return SpeedEstimateResult.Failure(
                    SpeedEstimateError.TimestampOutOfBounds("Query time is before vehicle state")
            )
        }

        val currentDist = status.scheduledDistanceAlongTrip
                ?: return SpeedEstimateResult.Failure(
                        SpeedEstimateError.InsufficientData("No scheduled distance along trip"))

        val speed = schedule?.speedAtDistance(currentDist)
                ?: return SpeedEstimateResult.Failure(
                        SpeedEstimateError.InsufficientData("Cannot compute schedule speed"))

        return SpeedEstimateResult.Success(DiracDistribution(speed))
    }
}
